repo2docker/buildpacks/r.py

```diff
@@ -55,6 +55,37 @@ def runtime(self):
 
         return self._runtime
 
+    @property
+    def r_version(self):
+        """Detect the R version for a given `runtime.txt`
+
+        Will return the version specified by the user or the current default
+        version.
+        """
+        version_map = {
+            "3.5": "3.5.3-1bionic",
+            "3.5.0": "3.5.0-1bionic",
+            "3.5.1": "3.5.1-2bionic",
+            "3.5.2": "3.5.2-1bionic",
+            "3.5.3": "3.5.3-1bionic",
+            "3.6": "3.6.1-3bionic",
+            "3.6.0": "3.6.0-2bionic",
+            "3.6.1": "3.6.1-3bionic",
+        }
+        # the default if nothing is specified
+        r_version = "3.6"
+
+        if not hasattr(self, "_r_version"):
+            parts = self.runtime.split("-")
+            if len(parts) == 5:
+                r_version = parts[1]
+
+            # see if we need to translate the specified version into a
+            # "full" version string
+            self._r_version = version_map.get(r_version, r_version)
```

Comment:
I think I would have an easier time reading this as `version_map.get(r_version, default = r_version)`.

What about the following test, which fails with the current code:

```python

def test_fallback_version_completion(tmpdir):
    tmpdir.chdir()

    with open("runtime.txt", "w") as f:
        f.write(f"r-3.8-2019-01-01")
    r = buildpacks.RBuildPack()
    assert r.r_version == "3.6.1-3bionic"

    with open("runtime.txt", "w") as f:
        f.write(f"r-2.15-2019-01-01")
    r = buildpacks.RBuildPack()
    assert r.r_version == "3.6.1-3bionic"
```

Do we want to capture the case where someone specifies an unsupported version and go to the default? We probably should fail then, right? It was just confusing to me to because `version.map(..)` will not return a full version string if there is no runtime version specified.

Reply:
I made it so that if the user specifies a version that isn't supported an exception is raised. I think that is what should happen in this case as my guess would be that the user made a typo when there is a version specified but it isn't a valid one.